lib/membrane_webm/serializer/webm.ex

```diff
@@ -0,0 +1,204 @@
+  defp construct_ebml_header() do
+    {
+      :EBML,
+      [
+        DocTypeReadVersion: 2,
+        DocTypeVersion: 4,
+        DocType: "webm",
+        EBMLMaxSizeLength: 8,
+        EBMLMaxIDLength: 4,
+        EBMLReadVersion: 1,
+        EBMLVersion: 1
+      ]
+    }
+  end
+
+  @spec construct_tracks(list({non_neg_integer, map})) :: {atom, list}
+  defp construct_tracks(tracks) do
+    {:Tracks, Enum.map(tracks, &construct_track_entry/1)}
+  end
+
+  defp construct_track_entry({id, %{caps: %Opus{channels: channels}, track_number: track_number}}) do
+    {:TrackEntry,
+     [
+       # this field will become important if membrane expands support for opus to more than 2 channels
+       CodecPrivate: Codecs.construct_opus_id_header(channels),
+       Audio: [
+         Channels: channels
+       ],
+       # 2 for audio
+       TrackType: 2,
+       CodecID: "A_OPUS",
+       TrackUID: id,
+       TrackNumber: track_number
+     ]}
+  end
+
+  defp construct_track_entry(
+         {id, %{caps: %VP8{width: width, height: height}, track_number: track_number}}
+       ) do
+    {:TrackEntry,
+     [
+       Video: [
+         PixelHeight: width,
+         PixelWidth: height
+       ],
+       # 1 for video
+       TrackType: 1,
+       CodecID: "V_VP8",
+       TrackUID: id,
+       TrackNumber: track_number
+     ]}
+  end
+
+  defp construct_track_entry(
+         {id, %{caps: %VP9{width: width, height: height}, track_number: track_number}}
+       ) do
+    {:TrackEntry,
+     [
+       Video: [
+         PixelHeight: width,
+         PixelWidth: height
+       ],
+       # 1 for video
+       TrackType: 1,
+       CodecID: "V_VP9",
+       TrackUID: id,
+       TrackNumber: track_number
+     ]}
+  end
+
+  # explanation of segment position:
+  # https://www.ietf.org/archive/id/draft-ietf-cellar-matroska-08.html#name-segment-position
+  @spec construct_seek_head(list({atom, list})) :: {atom, list}
+  defp construct_seek_head(elements) do
+    seeks =
+      elements
+      |> Enum.reduce({[], @seekhead_bytes + 1}, fn element, acc ->
+        {name, data} = element
+        {results, offset} = acc
+        new_acc = [{name, offset} | results]
+        new_offset = offset + byte_size(Helper.serialize({name, data}))
+        {new_acc, new_offset}
+      end)
+      |> elem(0)
+      |> Enum.map(fn {name, offset} ->
+        {:Seek,
+         [
+           {:SeekPosition, offset},
+           {:SeekID, EBML.encode_element_id(name)}
+         ]}
+      end)
+
+    {:SeekHead, seeks}
+  end
+
+  @spec construct_void({atom, list}) :: {atom, non_neg_integer}
+  defp construct_void(seek_head) do
+    void_width = @seekhead_bytes - byte_size(Helper.serialize(seek_head))
+    {:Void, void_width}
+  end
+
+  # this element MUST exist - because of TimestampScale
+  @spec construct_info() :: {atom, list}
+  defp construct_info() do
+    {
+      :Info,
+      [
+        # TODO: calculate Duration dynamically
+        Duration: 0,
```

Comment:
What about this?

Reply:
The `Info` element is saved at the very beginning of the file but the duration can only be known when the whole file is read. It is a similar problem with the Cues element. In fact the entire beginning of the webm file needs to be replaced by an updated header (of variable size) so that's a todo for now.